apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabaseBase.cs

```diff
@@ -0,0 +1,224 @@
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Api.Gax.ResourceNames;
+using Google.Cloud.Spanner.Admin.Database.V1;
+using Google.Cloud.Spanner.Admin.Instance.V1;
+using Google.Cloud.Spanner.Common.V1;
+using Google.Cloud.Spanner.V1.Internal.Logging;
+using Grpc.Core;
+using System;
+
+namespace Google.Cloud.Spanner.Data.CommonTesting;
+
+/// <summary>
+/// Base class for Spanner Test database.
+/// </summary>
+public abstract class SpannerTestDatabaseBase
+{
+    protected static readonly object s_lock = new object();
+
+    /// <summary>
+    /// The Spanner Host name to connect to. It is read from the environment variable "TEST_SPANNER_HOST".
+    /// </summary>
+    public virtual string SpannerHost { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_HOST", null);
+
+    /// <summary>
+    /// The Spanner port number to connect to. It is read from the environment variable "TEST_SPANNER_PORT".
+    /// </summary>
+    public virtual string SpannerPort { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_PORT", null);
+
+    /// <summary>
+    /// The Spanner Instance name to connect to. It is read from the environment variable "TEST_SPANNER_INSTANCE".
+    /// </summary>
+    public virtual string SpannerInstance { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_INSTANCE", "spannerintegration");
+
+    /// <summary>
+    /// The Spanner Database name to connect to. It is read from the environment variable "TEST_SPANNER_DATABASE".
+    /// The derived class must provide its implementation to return the database name.
+    /// </summary>
+    public abstract string SpannerDatabase { get; }
+
+    /// <summary>
+    /// Returns true if the database was created just for this test, or false if the database was an existing one
+    /// specified through an environment variable.
+    /// </summary>
+    public abstract bool Fresh { get; }
+
+    /// <summary>
+    /// Connection string including database.
+    /// </summary>
+    public string ConnectionString { get; }
+
+    /// <summary>
+    /// Connection string without the database.
+    /// </summary>
+    public string NoDbConnectionString { get; }
+
+    /// <summary>
+    /// The GCP Project ID to be used for the test database.
+    /// </summary>
+    public string ProjectId { get; }
+
+    /// <summary>
+    /// The fully-qualified database name.
+    /// </summary>
+    public DatabaseName DatabaseName { get; }
+
+    internal SpannerClientCreationOptions SpannerClientCreationOptions { get; }
+
+    protected SpannerTestDatabaseBase(string projectId, DatabaseDialect dialect)
+    {
+        TestLogger.Install();
+        ProjectId = projectId;
+        var builder = new SpannerConnectionStringBuilder
+        {
+            Host = SpannerHost,
+            DataSource = $"projects/{ProjectId}/instances/{SpannerInstance}",
+            EmulatorDetection = dialect == DatabaseDialect.Postgresql
```

Comment:
You have inheritance based on whether the dialect is PS or not. You shouldn't be needing to ask these questions here.

This constructor should receive a value for EmulatorDetection, not for the dialect.